### PR TITLE
Feature/bosch locator bridge utils

### PR DESCRIPTION
**Pull request by slaible**
This PR adds the package `bosch_locator_bridge_utils` to the repo. It is only tested with ROS Galactic right now.

### CODE REVIEW

File: bosch_locator_bridge_utils/launch/localization_launch.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            condition=IfCondition(LaunchConfiguration('use_locator_map'))),
+
+        Node(
+            package='nav2_map_server',
+            executable='map_server',
+            name='map_server',
+            output='screen',
+            parameters=[configured_params],
+            remappings=remappings,
+            condition=IfCondition(PythonExpression(['not ', LaunchConfiguration('use_locator_map')]))),
```
Comment (kfabian) [inline bosch_locator_bridge_utils/launch/localization_launch.py]:
```suggestion
            condition=UnlessCondition(LaunchConfiguration('use_locator_map'))),
```

File: bosch_locator_bridge_utils/launch/localization_launch.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+from ament_index_python.packages import get_package_share_directory
+
+from launch import LaunchDescription
+from launch.actions import DeclareLaunchArgument, SetEnvironmentVariable, IncludeLaunchDescription
+from launch.conditions import IfCondition
```
Comment (kfabian) [inline bosch_locator_bridge_utils/launch/localization_launch.py]:
```suggestion
from launch.conditions import IfCondition, UnlessCondition
```

File: bosch_locator_bridge_utils/launch/localization_launch.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+from launch.launch_description_sources import FrontendLaunchDescriptionSource
+from launch.substitutions import LaunchConfiguration, PythonExpression
```
Comment (kfabian) [inline bosch_locator_bridge_utils/launch/localization_launch.py]:
```suggestion
from launch.substitutions import LaunchConfiguration
```

File: bosch_locator_bridge_utils/params/params.yaml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+locator_node:
+  ros__parameters:
+    use_sim_time: False
+    base_frame_id: "base_footprint"
```
Comment (kfabian) [inline bosch_locator_bridge_utils/params/params.yaml:4]:
```suggestion
    base_frame_id: "base_link"
```
I think base_link would be a more generic default value.

File: bosch_locator_bridge_utils/config/localization.rviz [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+    - Class: rviz_default_plugins/TF
+      Enabled: true
+      Frame Timeout: 15
+      Frames:
+        All Enabled: true
+        base_footprint:
+          Value: true
+        base_link:
+          Value: true
+        caster_back_left_base_link:
```
Comment (kfabian) [inline bosch_locator_bridge_utils/config/localization.rviz]: Could these frames be removed?

**Comment by kfabian** [comment]
Since this a new package we could add a changelog file.

File: bosch_locator_bridge_utils/package.xml [hunk 6]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0"?>
+<package format="3">
+  <name>bosch_locator_bridge_utils</name>
+  <version>0.0.0</version>
```
Comment (kfabian) [inline bosch_locator_bridge_utils/package.xml:4]:
```suggestion
  <version>1.0.0</version>
```

**Review by kfabian — APPROVED** [review]
Very useful addition to this repo!

**Comment by kfabian** [comment]
I think we also need to add this package to the github workflow.

**Comment by slaible** [comment]
> Since this a new package we could add a changelog file.

The changelog file is automatically created with `catkin_generate_changelog` when preparing for the first release.

File: bosch_locator_bridge_utils/params/params.yaml (hunk 4 above)
Comment (slaible) [reply bosch_locator_bridge_utils/params/params.yaml:4]: Actually the 2D localization computes the pose of **base_footprint** relative to the map. When setting base_link as the base frame, the robot would sink a bit into the ground in RViz, if the z-position of base_link is greater than zero.

File: bosch_locator_bridge_utils/config/localization.rviz (hunk 5 above)
Comment (slaible) [reply bosch_locator_bridge_utils/config/localization.rviz]: Yes, just committed it.

File: bosch_locator_bridge_utils/package.xml (hunk 6 above)
Comment (slaible) [reply bosch_locator_bridge_utils/package.xml:4]: Not sure about the version numbers yet. The version numbers for the `bosch_locator_bridge` package are:
- 1.0.x for noetic
- 2.0.x for foxy
- 2.1.x for main

**Comment by slaible** [comment]
> I think we also need to add this package to the github workflow.

I added an extra workflow for this.

**Comment by ralph-lange** [comment]
My review comes much too late - but LGTM.

A remark on the workflow files: If you want to merge them for the different ROS distributions and Ubuntu versions into one, you can take https://github.com/boschresearch/fmi_adapter/blob/master/.github/workflows/build_and_test.yml#L16 as blueprint for how to build a list with two corresponding fields.

**Comment by slaible** [comment]
Thanks for the tip, but then we would not have a badge for each individual case.

**Comment by kfabian** [comment]
> My review comes much too late - but LGTM.
> 
> A remark on the workflow files: If you want to merge them for the different ROS distributions and Ubuntu versions into one, you can take https://github.com/boschresearch/fmi_adapter/blob/master/.github/workflows/build_and_test.yml#L16 as blueprint for how to build a list with two corresponding fields.

Also the checkout step is not needed, the action-ros-ci step will checkout the code again and the cron trigger works only for the default branch (noetic in our case).

**Comment by slaible** [comment]
@kfabian, for clarification: So we can remove the `actions/checkout@v2` step for all branches, and the `cron: '33 6 * * *'` trigger for all non-noetic branches?

**Comment by kfabian** [comment]
Yes we can remove `actions/checkout@v2` for all branches. The `cron: '33 6 * * *'` trigger has no effect on the non default branches, but maybe we can keep it for the future. To have daily builds for all ROS distros we would need to have the workflow files for each distro in the default branch, but the `ros-tooling/action-ros-ci` step always uses the branch it was called from, so this is not working currently. I have created an [issue](https://github.com/ros-tooling/action-ros-ci/issues/772) for `ros-tooling/action-ros-ci` to be able to checkout a specific branch.

**Pull request merged**